### ISSUE TITLE
Cannot use Rust trait exports without library mode in Kotlin

### ISSUE THREAD

**Issue by paxbun** [mozilla/uniffi-rs#1883]
Without library mode, UniFFI 0.25.2 pushes the trait information **twice** to `ComponentInterface`, one in `ComponentInterface::from_metadata`, and the other one in `macro_metadata::add_to_ci_from_library`. This makes duplicate overloads of methods like `*_uniffi_trait_eq_eq` or `*_uniffi_trait_debug` in generate bindings. This is trivial when using Python, but it makes compilation errors when using Kotlin or Swift.

To reproduce this issue, I modified `fixtures/trait-methods/tests/test_generated_bindings.rs` and `uniffi_bindgen/src/bindings/python/test.rs` as follows:

```rust
// fixtures/trait-methods/tests/test_generated_bindings.rs
#[test]
fn uniffi_foreign_language_testcase_test_py() -> uniffi::deps::anyhow::Result<()> {
    // use a directory that can be easily opened from the editor or the system explorer
    let tmp_dir = concat!(std::env!("CARGO_MANIFEST_DIR"), "/tests/tmp");
    std::fs::create_dir_all(tmp_dir).unwrap();
    uniffi::python_run_test(
        // std::env!("CARGO_TARGET_TMPDIR"),
        tmp_dir,
        std::env!("CARGO_PKG_NAME"),
        concat!(std::env!("CARGO_MANIFEST_DIR"), "/tests/bindings/test.py"),
    )
}
```

```rust
// uniffi_bindgen/src/bindings/python/test.rs
...
    // generate_bindings(
    //     &cdylib_path,
    //     None,
    //     &[TargetLanguage::Python],
    //     &out_dir,
    //     false,
    // )?;
    crate::generate_bindings(
        Utf8Path::new(concat!(
            std::env!("CARGO_MANIFEST_DIR"),
            "/../fixtures/trait-methods/src/trait_methods.udl"
        )),
        None,
        vec![TargetLanguage::Python],
        Some(&out_dir),
        Some(&cdylib_path),
        None,
        false,
    )?;
...
```

I was implementing Rust trait exports for `uniffi-kotlin-multiplatform-bindings` but encountered this issue. Please see https://gitlab.com/trixnity/uniffi-kotlin-multiplatform-bindings/-/issues/28 for more details. We're bypassing this issue by using library mode in our unit tests, but I think it needs to be fixed in the original UniFFI repo.

**Comment by mhammond** [comment]
We'd probably take a fix for this, but the direction we are heading is to only support library mode, as that's the only way "external types" might work.

**Comment by ptitjes** [comment]
> We'd probably take a fix for this, but the direction we are heading is to only support library mode, as that's the only way "external types" might work.

Do you mean that there would be only one workflow for the user? That is:
1. use the scaffolding generator, that reads the UDL file and generates a library with metadata included, and then
2. use the foreign language binding generator that reads the metadata from the library.

Did I understand correctly? If yes do you know when the other workflow would be dropped?
(Asking this to plan ahead for our Kotlin Multiplatform bindings and its Gradle plugin...)

**Comment by mhammond** [comment]
Yeah, that is what I meant - although part of (1) is already optional because UDL files are optional. (2) is necessary so we can "resolve" external types used by the crate or types exposed via procmacros (note that loading the UDL file when generating foreign bindings still isn't going to help you learn about types exposed via procmacros)

I'm not sure we have decided to actually drop support for the other workflow though, especially while it continues to work OK for users who aren't trying to use external types or procmacros - which is why I was suggesting we would take a patch. IIUC, library mode is used exclusively inside both the uniffi-rs repo *and* inside all other mozilla repos which use uniffi.

**Comment by ptitjes** [comment]
@mhammond `uniffi_macros::build_foreign_language_testcases` (which is used in the test fixtures) always use `uniffi_bindgen::library_mode::generate_bindings`. What would you prefer in order to test a fix to this issue that would use `uniffi_bindgen::generate_external_bindings`?

The best option I see for now is to add a `build_foreign_language_testcases_from_udl` function that would pass `true` to a new `from_udl: bool` parameter down to the `run_script` function calls. We would then selectively apply `build_foreign_language_testcases_from_udl` in addition to current `build_foreign_language_testcases`. Here I would only apply it to the `trait-methods` fixture.

What do you think?

**Comment by mhammond** [comment]
I assume the fix will be such that there's probably no real advantage/requirement to have tests in each of the bindings - indeed, it might even be such that a pure rust test is all we need. But assuming we want just one binding, I'd say we go with Python and hack together a solution for just that 1 crate. If that turns out to be too difficult etc, then what you suggest sounds fine.

**Comment by ptitjes** [comment]
@mhammond In fact, looking at the code, I don't clearly see what the fix would be...

I mean, in `uniffi_bingen::generate_external_bindings` we have those four consecutive lines:

```
    let mut component = parse_udl(udl_file.as_ref(), &crate_name)?;
    if let Some(ref library_file) = library_file {
        macro_metadata::add_to_ci_from_library(&mut component, library_file.as_ref())?;
    }
```

The first one gets metadata by parsing the UDL and the other three lines get metadata from the library.

Assuming the metadata in the library has been produced from both the UDL and the proc-macros, the only possible fix is to not get metadata from parsing the UDL.

**Comment by mhammond** [comment]
That seems correct for the current implementation, yeah. Isn't it getting easier to use library-mode?

**Comment by mhammond** [comment]
Sorry I didn't look closer before and that I focused on library mode, but yeah, this a bug in uniffi - we shouldn't end up with dupe traitmethods.
